File: samples/python/src/phxp/protocol.py
from __future__ import annotations

import struct
from collections.abc import Callable
from dataclasses import dataclass
from enum import IntEnum

MAGIC = b"PHXP"
VERSION = 1
HEADER_LENGTH = 40
MAX_PACKET_LENGTH = 512
MAX_SNI_LENGTH = 253
_HEADER = struct.Struct("!4sBBH16sIQHH")
_ZERO_ID = bytes(16)


class ProtocolError(ValueError):
    pass


class MessageType(IntEnum):
    HELLO = 1
    READY = 2
    HANDOFF = 3
    ADOPTED = 4
    REJECTED = 5


@dataclass(frozen=True, slots=True)
class Message:
    type: MessageType
    connection_id: bytes = _ZERO_ID
    peeked_length: int = 0
    accepted_at_ns: int = 0
    requested_sni: str = ""
    rejection_code: int = 0


def encode(message: Message) -> bytes:
    _validate_uint(message.peeked_length, 32, "peeked length")
    _validate_uint(message.accepted_at_ns, 64, "accepted timestamp")
    _validate_uint(message.rejection_code, 16, "rejection code")
    if len(message.connection_id) != 16:
        raise ProtocolError("PHXP connection ID must contain exactly 16 bytes")

    payload = b""
    if message.type in (MessageType.HELLO, MessageType.READY):
        if (
            message.connection_id != _ZERO_ID
            or message.peeked_length
            or message.accepted_at_ns
            or message.requested_sni
            or message.rejection_code
        ):
            raise ProtocolError("PHXP handshake has unexpected field values")
    elif message.type is MessageType.HANDOFF:
        try:
            payload = message.requested_sni.encode("utf-8")
        except UnicodeEncodeError as error:
            raise ProtocolError("PHXP handoff SNI is not valid UTF-8") from error
        if not 1 <= len(payload) <= MAX_SNI_LENGTH:
            raise ProtocolError("PHXP handoff SNI length is outside protocol bounds")
        if message.rejection_code:
            raise ProtocolError("PHXP handoff request has a rejection code")
    elif message.type is MessageType.ADOPTED:
        if (
            message.peeked_length
            or message.accepted_at_ns
            or message.requested_sni
            or message.rejection_code
        ):
            raise ProtocolError("PHXP adopted response has unexpected field values")
    elif message.type is MessageType.REJECTED:
        if (
            message.peeked_length
            or message.accepted_at_ns
            or message.requested_sni
            or not message.rejection_code
        ):
            raise ProtocolError("PHXP rejected response has unexpected field values")
    else:
        raise ProtocolError(f"unknown PHXP message type {message.type}")

    if HEADER_LENGTH + len(payload) > MAX_PACKET_LENGTH:
        raise ProtocolError("PHXP packet exceeds protocol limit")
    return (
        _HEADER.pack(
            MAGIC,
            VERSION,
            int(message.type),
            0,
            message.connection_id,
            message.peeked_length,
            message.accepted_at_ns,
            len(payload),
            message.rejection_code,
        )
        + payload
    )


def decode(packet: bytes) -> Message:
    length = frame_length(packet)
    if len(packet) != length:
        raise ProtocolError("PHXP payload length does not match packet")
    (
        _magic,
        _version,
        raw_type,
        _flags,
        connection_id,
        peeked_length,
        accepted_at_ns,
        payload_length,
        rejection_code,
    ) = _HEADER.unpack_from(packet)
    message_type = MessageType(raw_type)

    if message_type in (MessageType.HELLO, MessageType.READY):
        if (
            payload_length
            or connection_id != _ZERO_ID
            or peeked_length
            or accepted_at_ns
            or rejection_code
        ):
            raise ProtocolError("PHXP handshake has unexpected field values")
        return Message(message_type)
    if message_type is MessageType.HANDOFF:
        if not 1 <= payload_length <= MAX_SNI_LENGTH or rejection_code:
            raise ProtocolError("PHXP handoff request has invalid field values")
        try:
            requested_sni = packet[HEADER_LENGTH:].decode("utf-8")
        except UnicodeDecodeError as error:
            raise ProtocolError("PHXP handoff SNI is not valid UTF-8") from error
        return Message(
            message_type,
            connection_id,
            peeked_length,
            accepted_at_ns,
            requested_sni,
        )
    if payload_length or peeked_length or accepted_at_ns:
        raise ProtocolError("PHXP response has unexpected field values")
    if message_type is MessageType.ADOPTED:
        if rejection_code:
            raise ProtocolError("PHXP response has unexpected field values")
        return Message(message_type, connection_id)
    if not rejection_code:
        raise ProtocolError("PHXP rejection has invalid field values")
    return Message(message_type, connection_id, rejection_code=rejection_code)


def frame_length(header: bytes) -> int:
    if len(header) < HEADER_LENGTH:
        raise ProtocolError("PHXP packet is shorter than its fixed header")
    magic, version, raw_type, flags = struct.unpack_from("!4sBBH", header)
    if magic != MAGIC:
        raise ProtocolError("PHXP packet has invalid magic")
    if version != VERSION:
        raise ProtocolError(f"unsupported PHXP protocol version {version}")
    try:
        MessageType(raw_type)
    except ValueError as error:
        raise ProtocolError(f"unknown PHXP message type {raw_type}") from error
    if flags:
        raise ProtocolError("PHXP packet uses unsupported flags")
    payload_length = struct.unpack_from("!H", header, 36)[0]
    length = HEADER_LENGTH + payload_length
    if length > MAX_PACKET_LENGTH:
        raise ProtocolError("PHXP packet exceeds protocol limit")
    return length


def accumulate_stream_frame(initial: bytes, receive: Callable[[int], bytes]) -> bytes:
    if len(initial) > MAX_PACKET_LENGTH:
        raise ProtocolError("PHXP stream contains bytes beyond the maximum frame")
    frame = bytearray(initial)
    while len(frame) < HEADER_LENGTH:
        chunk = receive(HEADER_LENGTH - len(frame))
        if not chunk:
            raise ProtocolError("unexpected EOF in PHXP frame header")
        frame.extend(chunk)
    length = frame_length(frame)
    if len(frame) > length:
        raise ProtocolError("PHXP stream contains bytes beyond the declared frame")
    while len(frame) < length:
        chunk = receive(length - len(frame))
        if not chunk:
            raise ProtocolError("unexpected EOF in PHXP frame payload")
        frame.extend(chunk)
    return bytes(frame)


def _validate_uint(value: int, bits: int, field: str) -> None:
    if not isinstance(value, int) or not 0 <= value < 1 << bits:
        raise ProtocolError(f"PHXP {field} is outside its unsigned {bits}-bit range")

File: samples/go/phxp/protocol.go
package phxp

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"unicode/utf8"
)

const (
	Version         = 1
	HeaderLength    = 40
	MaxPacketLength = 512
	MaxSNILength    = 253
)

type MessageType uint8

const (
	TypeHello    MessageType = 1
	TypeReady    MessageType = 2
	TypeHandoff  MessageType = 3
	TypeAdopted  MessageType = 4
	TypeRejected MessageType = 5
)

type Message struct {
	Type          MessageType
	ConnectionID  [16]byte
	PeekedLength  uint32
	AcceptedAtNS  uint64
	RequestedSNI  string
	RejectionCode uint16
}

func Encode(message Message) ([]byte, error) {
	var payload []byte
	switch message.Type {
	case TypeHello, TypeReady:
		if message.ConnectionID != ([16]byte{}) || message.PeekedLength != 0 ||
			message.AcceptedAtNS != 0 || message.RequestedSNI != "" ||
			message.RejectionCode != 0 {
			return nil, errors.New("PHXP handshake has unexpected field values")
		}
	case TypeHandoff:
		payload = []byte(message.RequestedSNI)
		if len(payload) == 0 || len(payload) > MaxSNILength {
			return nil, errors.New("PHXP handoff SNI length is outside protocol bounds")
		}
		if !utf8.Valid(payload) {
			return nil, errors.New("PHXP handoff SNI is not valid UTF-8")
		}
		if message.RejectionCode != 0 {
			return nil, errors.New("PHXP handoff request has a rejection code")
		}
	case TypeAdopted:
		if message.PeekedLength != 0 || message.AcceptedAtNS != 0 ||
			message.RequestedSNI != "" || message.RejectionCode != 0 {
			return nil, errors.New("PHXP adopted response has unexpected field values")
		}
	case TypeRejected:
		if message.PeekedLength != 0 || message.AcceptedAtNS != 0 ||
			message.RequestedSNI != "" || message.RejectionCode == 0 {
			return nil, errors.New("PHXP rejected response has unexpected field values")
		}
	default:
		return nil, fmt.Errorf("unknown PHXP message type %d", message.Type)
	}

	packetLength := HeaderLength + len(payload)
	if packetLength > MaxPacketLength {
		return nil, errors.New("PHXP packet exceeds protocol limit")
	}
	packet := make([]byte, packetLength)
	copy(packet[0:4], "PHXP")
	packet[4] = Version
	packet[5] = byte(message.Type)
	copy(packet[8:24], message.ConnectionID[:])
	binary.BigEndian.PutUint32(packet[24:28], message.PeekedLength)
	binary.BigEndian.PutUint64(packet[28:36], message.AcceptedAtNS)
	binary.BigEndian.PutUint16(packet[36:38], uint16(len(payload)))
	binary.BigEndian.PutUint16(packet[38:40], message.RejectionCode)
	copy(packet[HeaderLength:], payload)
	return packet, nil
}

func Decode(packet []byte) (Message, error) {
	frameLength, err := FrameLength(packet)
	if err != nil {
		return Message{}, err
	}
	if len(packet) != frameLength {
		return Message{}, errors.New("PHXP payload length does not match packet")
	}

	var message Message
	message.Type = MessageType(packet[5])
	copy(message.ConnectionID[:], packet[8:24])
	message.PeekedLength = binary.BigEndian.Uint32(packet[24:28])
	message.AcceptedAtNS = binary.BigEndian.Uint64(packet[28:36])
	payloadLength := int(binary.BigEndian.Uint16(packet[36:38]))
	message.RejectionCode = binary.BigEndian.Uint16(packet[38:40])

	switch message.Type {
	case TypeHello, TypeReady:
		if payloadLength != 0 || message.ConnectionID != ([16]byte{}) ||
			message.PeekedLength != 0 || message.AcceptedAtNS != 0 ||
			message.RejectionCode != 0 {
			return Message{}, errors.New("PHXP handshake has unexpected field values")
		}
	case TypeHandoff:
		if payloadLength == 0 || payloadLength > MaxSNILength || message.RejectionCode != 0 {
			return Message{}, errors.New("PHXP handoff request has invalid field values")
		}
		payload := packet[HeaderLength:]
		if !utf8.Valid(payload) {
			return Message{}, errors.New("PHXP handoff SNI is not valid UTF-8")
		}
		message.RequestedSNI = string(payload)
	case TypeAdopted:
		if payloadLength != 0 || message.PeekedLength != 0 ||
			message.AcceptedAtNS != 0 || message.RejectionCode != 0 {
			return Message{}, errors.New("PHXP response has unexpected field values")
		}
	case TypeRejected:
		if payloadLength != 0 || message.PeekedLength != 0 ||
			message.AcceptedAtNS != 0 || message.RejectionCode == 0 {
			return Message{}, errors.New("PHXP rejection has invalid field values")
		}
	default:
		return Message{}, fmt.Errorf("unknown PHXP message type %d", message.Type)
	}
	return message, nil
}

func FrameLength(header []byte) (int, error) {
	if len(header) < HeaderLength {
		return 0, errors.New("PHXP packet is shorter than its fixed header")
	}
	if !bytes.Equal(header[0:4], []byte("PHXP")) {
		return 0, errors.New("PHXP packet has invalid magic")
	}
	if header[4] != Version {
		return 0, fmt.Errorf("unsupported PHXP protocol version %d", header[4])
	}
	switch MessageType(header[5]) {
	case TypeHello, TypeReady, TypeHandoff, TypeAdopted, TypeRejected:
	default:
		return 0, fmt.Errorf("unknown PHXP message type %d", header[5])
	}
	if header[6] != 0 || header[7] != 0 {
		return 0, errors.New("PHXP packet uses unsupported flags")
	}
	length := HeaderLength + int(binary.BigEndian.Uint16(header[36:38]))
	if length > MaxPacketLength {
		return 0, errors.New("PHXP packet exceeds protocol limit")
	}
	return length, nil
}

func readStreamFrame(reader io.Reader, initial []byte) ([]byte, error) {
	frame := append([]byte(nil), initial...)
	if len(frame) > MaxPacketLength {
		return nil, errors.New("PHXP stream contains bytes beyond the maximum frame")
	}
	if len(frame) < HeaderLength {
		missing := HeaderLength - len(frame)
		frame = append(frame, make([]byte, missing)...)
		if _, err := io.ReadFull(reader, frame[len(frame)-missing:]); err != nil {
			return nil, fmt.Errorf("unexpected EOF in PHXP frame header: %w", err)
		}
	}
	length, err := FrameLength(frame[:HeaderLength])
	if err != nil {
		return nil, err
	}
	if len(frame) > length {
		return nil, errors.New("PHXP stream contains bytes beyond the declared frame")
	}
	if len(frame) < length {
		missing := length - len(frame)
		frame = append(frame, make([]byte, missing)...)
		if _, err := io.ReadFull(reader, frame[len(frame)-missing:]); err != nil {
			return nil, fmt.Errorf("unexpected EOF in PHXP frame payload: %w", err)
		}
	}
	return frame, nil
}
